src/components/input/input.ts

```diff
@@ -0,0 +1,227 @@
+
+
+/**
+ * The hint directive, used to tag content as hint labels (going under the input).
+ */
+@Directive({
+  selector: 'md-hint',
+  host: {
+    '[class.md-right]': 'align == "end"',
+    '[class.md-hint]': 'true'
+  }
+})
+export class MdHint {
+  @Input() @OneOf(['start', 'end']) align: string;
+}
+
+
+/**
+ * <md-input> component.
+ */
+@Component({
+  selector: 'md-input',
+  templateUrl: 'components/input/input.html',
+  styleUrls: ['components/input/input.css'],
+  providers: [MdAria, MD_INPUT_CONTROL_VALUE_ACCESSOR],
+  directives: [MdAriaTarget],
+})
+export class MdInput implements ControlValueAccessor, AfterContentInit, OnChanges {
+  private _focused: boolean = false;
+  private _onTouched: () => void = noop;
+  private _onChangeFn: (_: any) => void = noop;
+  private _value: any = '';
+
+  /**
+   * Aria related inputs.
+   */
+  @Input('aria-label') ariaLabel: string;
+  @Input('aria-labelledby') ariaLabelledBy: string;
+  @Input('aria-disabled') @BooleanFieldValue() ariaDisabled: boolean;
+  @Input('aria-required') @BooleanFieldValue() ariaRequired: boolean;
+  @Input('aria-invalid') @BooleanFieldValue() ariaInvalid: boolean;
+
+  /**
+   * Content directives.
+   */
+  @ContentChild(MdPlaceholder) private _placeholderChild: MdPlaceholder;
+  @ContentChildren(MdHint) private _hintChildren: QueryList<MdHint>;
+
+  /**
+   * Readonly properties.
+   */
+  get focused() { return this._focused; }
+  get empty() { return this._value == null || this._value === ''; }
+  get characterCount(): number {
+    return this.empty ? 0 : ('' + this._value).length;
+  }
+
+  /**
+   * Bindings.
+   */
+  @Input() @OneOf(['start', 'end']) align: string = 'start';
+  @Input() @BooleanFieldValue() disabled: boolean = false;
+  @Input() @OneOf(['primary', 'accent']) dividerColor: string = 'primary';
+  @Input() @BooleanFieldValue() floatingPlaceholder: boolean = true;
+  @Input() hintLabel: string = '';
+  @Input() id: string = `md-input-${nextUniqueId++}`;
+  @Input() maxLength: number = -1;
+  @Input() placeholder: string;
+  @Input() @BooleanFieldValue() required: boolean = false;
+  @Input() type: string = 'text';
+
+  get value(): any { return this._value; };
+  @Input() set value(v: any) {
+    if (this._setValue(v)) {
+      this._onChangeFn(v);
+    }
+  }
+
+  // This is to remove the `align` property of the `md-input` itself. Otherwise HTML5
+  // might place it as RTL when we don't want to. We still want to use `align` as an
+  // Input though, so we use HostBinding.
+  @HostBinding('attr.align') private get _align(): any { return null; }
+
+  onFocus() {
+    this._focused = true;
+  }
+  onBlur() {
+    this._focused = false;
+    this._onTouched();
+  }
+
+  hasPlaceholder(): boolean {
+    return this.placeholder != null || this._placeholderChild != null;
+  }
+
+  private _setValue(value: any): boolean {
+    if (this._value !== value) {
+      this._value = value;
+      return true;
+    } else {
+      return false;
+    }
```

Comment:
Unfortunately I cannot do that. There are cases where we want to change the `_value` but not trigger the event; if we do that on a call to `writeValue()` for example, it results in an infinite loop.